File: Go/evaluator/evaluator.go
package evaluator

import (
	"CornyLang/ast"
	"CornyLang/object"
	"CornyLang/token"
	"fmt"
)

var TRUE = &object.BooleanObj{Value: true}
var FALSE = &object.BooleanObj{Value: false}
var NULL = &object.NullObj{}

/**
* Helpers functions
 */
func isError(obj object.Object) bool {
	return obj != nil && obj.Type() == object.ERROR_OBJ
}

func evalExpressions(exps []ast.Expression, env *object.Environment) []object.Object {
	var result []object.Object

	for _, e := range exps {
		evaluated := Eval(e, env)
		if isError(evaluated) {
			return []object.Object{evaluated}
		}
		result = append(result, evaluated)
	}

	return result
}

func Eval(node ast.Node, env *object.Environment) object.Object {
	switch node := node.(type) {
	case *ast.ProgramNode:
		return evalProgram(node, env)
	case *ast.BlockStmtNode:
		return evalBlock(node, env)
	case *ast.ExpressionStatement:
		return Eval(node.Expression, env)
	case *ast.NumberNode:
		return &object.NumberObj{Value: node.Value}
	case *ast.StringNode:
		return &object.StringObj{Value: node.Value}
	case *ast.BooleanNode:
		return evalBooleanNode(node)
	case *ast.NullNode:
		return NULL
	case *ast.FunctionLiteralNode:
		return evalFunctionLiteralNode(node, env)
	case *ast.ArrayLiteralNode:
		return evalArrayLiteral(node, env)
	case *ast.HashLiteralNode:
		return evalHashLiteral(node, env)
	case *ast.ReturnNode:
		return Eval(node.Value, env)
	case *ast.LetNode:
		return evalLet(node, env)
	case *ast.IfExprNode:
		return evalIfExpr(node, env)
	case *ast.IdentifierNode:
		return evalIdentifier(node, env)
	case *ast.CallExprNode:
		return evalCallExpr(node, env)
	case *ast.UnaryNode:
		return evalUnaryExpr(node, env)
	case *ast.BinOpNode:
		return evalBinaryExpr(node, env)
	default:
		return nil
	}
}

func evalProgram(program *ast.ProgramNode, env *object.Environment) object.Object {
	var result object.Object

	for _, statememt := range program.Statements {
		result = Eval(statememt, env)
		switch result := result.(type) {
		case *object.ReturnObj:
			return result.Value // return the value object.
		case *object.ErrorObj:
			return result // return the error object.
		}
	}

	return result
}

func evalBlock(block *ast.BlockStmtNode, env *object.Environment) object.Object {
	var result object.Object

	for _, statement := range block.Statements {
		result = Eval(statement, env)
		if result != nil {
			if result.Type() == object.ERROR_OBJ || result.Type() == object.RETURN_OBJ {
				return result
			}
		}
	}

	return result
}

func evalLet(letNode *ast.LetNode, env *object.Environment) object.Object {
	var valueObj = Eval(letNode.Value, env)
	if isError(valueObj) {
		return valueObj
	}
	// register in symbol table
	env.Set(letNode.Name.Value, valueObj)

	return valueObj
}

// evalIfExpr
func evalIfExpr(ifNode *ast.IfExprNode, env *object.Environment) object.Object {
	var conditionObj = Eval(ifNode.Condition, env)
	if isError(conditionObj) {
		return conditionObj
	}
	boolConditionObj, ok := conditionObj.(*object.BooleanObj)
	if !ok {
		return object.NewError("Invalid data type for if condition.\n")
	}
	if boolConditionObj.Value {
		return Eval(ifNode.Consequence, env)
	} else {
		if ifNode.Alternative != nil {
			return Eval(ifNode.Alternative, env)
		} else {
			return NULL
		}
	}
}

// evalCallExpr
func evalCallExpr(callExpr *ast.CallExprNode, env *object.Environment) object.Object {
	// eval the callee
	var calleeObj = Eval(callExpr.Callee, env)

	if isError(calleeObj) {
		return calleeObj
	}

	var args []object.Object
	if callExpr.Token.Type != token.TT_DOT {
		// eval arguments
		args = evalExpressions(callExpr.Arguments, env)
		if len(args) == 1 && isError(args[0]) {
			return args[0]
		}
	} else {
		if len(callExpr.Arguments) != 1 {
			return object.NewError("wrong number of arguments for builtin function.")
		}
		if callExprNode, ok := callExpr.Arguments[0].(*ast.CallExprNode); ok {
			// check for identifierNode type
			if identifierNode, ok := callExprNode.Callee.(*ast.IdentifierNode); ok {
				// identifierNode must be resolved into builtinObj
				var builtinObj = evalIdentifier(identifierNode, env)
				if builtinObj, ok := builtinObj.(*object.BuiltinObj); ok {
					// this builtin function takes extra arguments?
					if !builtinObj.ExtraArgs && len(callExprNode.Arguments) > 0 {
						return object.NewError("Unexpected arguments.")
					}
					var newArgs = []object.Object{calleeObj} // we always pass the callee reference.
					if builtinObj.ExtraArgs {
						// pass extra arguments gathered by `callExpr.Arguments`
						var evaluatedArgs = evalExpressions(callExprNode.Arguments, env)
						newArgs = append(newArgs, evaluatedArgs...)
					}
					// execute the builtin function and get the result
					var resultObj = builtinObj.Func(newArgs...)
					// get the identifier from callExpr.Callee
					if identifierNode, ok := callExpr.Callee.(*ast.IdentifierNode); ok {
						env.Set(identifierNode.Value, resultObj)
					}
					return resultObj

				}
			}
		}
		return object.NewError(fmt.Sprintf("Invalid argument for builtin function using the dot(.) semantic: %s\n", callExpr.Arguments[0].String()))
	}

	switch calleeObj := calleeObj.(type) {
	case *object.FunctionObj:
		return evalFunction(calleeObj, args)
	case *object.ArrayObj:
		if callExpr.Token.Type == token.TT_LBRACKET {
			return evalArrayAccess(calleeObj, args)
		}
	case *object.HashObj:
		if callExpr.Token.Type == token.TT_LBRACKET {
			return evalHashAccess(calleeObj, args)
		}
	case *object.StringObj:
		if callExpr.Token.Type == token.TT_LBRACKET {
			return evalStringAccess(calleeObj, args)
		} else if callExpr.Token.Type == token.TT_DOT {
			return NULL
		}
	case *object.BuiltinObj: // NOTE: i'm not sure if this is necessary
		return calleeObj.Func(args...)
	default:
		//return object.NewError(fmt.Sprintf("not a function: %s\n", calleeObj.Type()))
		return calleeObj // TODO: this must be checked deeper.
	}
	return nil
}

func evalFunction(functionObj *object.FunctionObj, args []object.Object) object.Object {
	// 1. create new environment and close it with FunctionObj.Env
	var newEnv = object.NewEnvironment()
	newEnv.Parent = functionObj.Env

	// 2. check for function arity
	var numArgs = len(args)
	var numParams = len(functionObj.Parameters)
	if numArgs != numParams {
		return object.NewError(fmt.Sprintf("Unexpected arguments, got: %d, want: %d\n", numArgs, numParams))
	}
	// 3. fill the new environment with arguments
	for paramIdx, param := range functionObj.Parameters {
		newEnv.Set(param.Value, args[paramIdx])
	}
	// 4. execute the function with new environment
	var resultObj = Eval(functionObj.Body, newEnv)
	if isError(resultObj) {
		return resultObj
	}
	// check for return object
	if resultObj, ok := resultObj.(*object.ReturnObj); ok {
		return resultObj.Value
	}

	return resultObj
}

// evalArrayAccess
func evalArrayAccess(arrayObj *object.ArrayObj, args []object.Object) object.Object {
	indexObj, ok := args[0].(*object.NumberObj)
	if !ok {
		return object.NewError("Invalid subscript reference.")
	}
	// check for out of bounds
	var index = int(indexObj.Value)
	if index < 0 || index >= len(arrayObj.Elements) {
		return object.NewError("Index out of bounds.")
	}

	return arrayObj.Elements[index]
}

// evalHashAccess
func evalHashAccess(hashObj *object.HashObj, args []object.Object) object.Object {
	indexObj, ok := args[0].(*object.StringObj)
	if !ok {
		return object.NewError("Invalid subscript reference")
	}
	var key = indexObj.Value
	valueObj, ok := hashObj.Elements[key]
	if !ok {
		return NULL
	}
	return valueObj
}

// evalStringAccess
func evalStringAccess(stringObj *object.StringObj, args []object.Object) object.Object {
	indexObj, ok := args[0].(*object.NumberObj)
	if !ok {
		return object.NewError("Invalid subscript reference")
	}
	var index = int(indexObj.Value)
	if index < 0 || index >= len(stringObj.Value) {
		return object.NewError("Index out of bounds")
	}

	return &object.StringObj{Value: string(stringObj.Value[index])}
}

// evalArrayLiteral
func evalArrayLiteral(arrayNode *ast.ArrayLiteralNode, env *object.Environment) object.Object {
	var arrayObj = &object.ArrayObj{}

	// evaluate the elements of the array
	if len(arrayNode.Elements) > 0 {
		arrayObj.Elements = evalExpressions(arrayNode.Elements, env)
	}

	return arrayObj
}

// evalHashLiteral
func evalHashLiteral(hashNode *ast.HashLiteralNode, env *object.Environment) object.Object {
	var hashObj = &object.HashObj{}

	// evaluate expressions from keys
	var keysObj = evalExpressions(hashNode.Keys, env)
	// evaluate expressions from values
	var valuesObj = evalExpressions(hashNode.Values, env)

	hashObj.Elements = make(map[string]object.Object)

	// fill the internal hashObj dictionary
	for i := 0; i < len(keysObj); i++ {
		keyStr, ok := keysObj[i].(*object.StringObj)
		if !ok {
			return object.NewError("Invalid data type for key.")
		}
		hashObj.Elements[keyStr.Value] = valuesObj[i]
	}

	return hashObj
}

// evalUnaryExpr
func evalUnaryExpr(unaryNode *ast.UnaryNode, env *object.Environment) object.Object {
	var resultObj object.Object
	var rightObj = Eval(unaryNode.Right, env)

	if isError(rightObj) {
		return rightObj
	}
	// check the token operator
	switch unaryNode.Operator.Type {
	case token.TT_MINUS:
		numberObj, ok := rightObj.(*object.NumberObj)
		if !ok {
			return object.NewError("Invalid data type.")
		}
		resultObj = &object.NumberObj{Value: numberObj.Value * -1}
	case token.TT_NOT:
		booleanObj, ok := rightObj.(*object.BooleanObj)
		if !ok {
			return object.NewError("Invalid data type.")
		}
		resultObj = &object.BooleanObj{Value: !booleanObj.Value}
	default:
		return object.NewError(fmt.Sprintf("Invalid operator: %s", unaryNode.Operator.Literal))
	}
	return resultObj
}

// evalBinaryExpr
func evalBinaryExpr(binOpNode *ast.BinOpNode, env *object.Environment) object.Object {
	if binOpNode.Operator.Type == token.TT_AND || binOpNode.Operator.Type == token.TT_OR {
		return evalLogicalExpression(binOpNode, env)
	}
	// we need to know both operands types
	var leftObj = Eval(binOpNode.Left, env)
	if isError(leftObj) {
		return leftObj
	}
	var rightObj = Eval(binOpNode.Right, env)
	if isError(rightObj) {
		return rightObj
	}
	// now based on their types we perform the correct operation
	if leftObj.Type() == object.NUMBER_OBJ && rightObj.Type() == object.NUMBER_OBJ {
		leftNumber, _ := leftObj.(*object.NumberObj)
		rightNumber, _ := rightObj.(*object.NumberObj)

		return evalBinaryInteger(leftNumber, binOpNode.Operator, rightNumber)
	}
	if leftObj.Type() == object.STRING_OBJ && rightObj.Type() == object.STRING_OBJ {
		leftString, _ := leftObj.(*object.StringObj)
		rightString, _ := rightObj.(*object.StringObj)
		return evalBinaryString(leftString, binOpNode.Operator, rightString)
	}
	if leftObj.Type() == object.BOOLEAN_OBJ && rightObj.Type() == object.BOOLEAN_OBJ {
		leftBoolean, _ := leftObj.(*object.BooleanObj)
		rightBoolean, _ := rightObj.(*object.BooleanObj)
		var (
			leftNumber  = &object.NumberObj{Value: 0}
			rightNumber = &object.NumberObj{Value: 0}
		)
		if leftBoolean.Value {
			leftNumber.Value = 1
		}
		if rightBoolean.Value {
			rightNumber.Value = 1
		}

		return evalBinaryInteger(leftNumber, binOpNode.Operator, rightNumber)
	}
	return object.NewError(fmt.Sprintf("Incompatible data types: %s, %s\n", leftObj.Type(), rightObj.Type()))
}

// evalLogicalExpression
func evalLogicalExpression(binOpNode *ast.BinOpNode, env *object.Environment) object.Object {
	// evaluate the left operator
	var leftObj = Eval(binOpNode.Left, env)
	if isError(leftObj) {
		return leftObj
	}
	leftBinObj, ok := leftObj.(*object.BooleanObj)
	if !ok {
		return object.NewError("Invalid left hand type operand.")
	}
	if binOpNode.Operator.Type == token.TT_AND {
		// if leftObj is false then we do nothing
		if !leftBinObj.Value {
			return FALSE
		}
		// otherwise we need to evaluate the right hand operand
		var rightObj = Eval(binOpNode.Right, env)
		if isError(rightObj) {
			return rightObj
		}
		rightBinOp, ok := rightObj.(*object.BooleanObj)
		if !ok {
			return object.NewError("Invalid right hand type operand.")
		}
		if rightBinOp.Value {
			return TRUE
		} else {
			return FALSE
		}
	} else if binOpNode.Operator.Type == token.TT_OR {
		// if leftObj is true then we do nothing
		if leftBinObj.Value {
			return TRUE
		}
		// otherwise we need to evaluate the right hand operand
		var rightObj = Eval(binOpNode.Right, env)
		if isError(rightObj) {
			return rightObj
		}
		rightBinOp, ok := rightObj.(*object.BooleanObj)
		if !ok {
			return object.NewError("Invalid right hand type operand.")
		}
		if rightBinOp.Value {
			return TRUE
		} else {
			return FALSE
		}
	}
	return object.NewError(fmt.Sprintf("Invalid operator for logical expression: %s\n", binOpNode.Operator.Type))
}

// evalBinaryInteger
func evalBinaryInteger(leftObj *object.NumberObj, ope token.Token, rightObj *object.NumberObj) object.Object {
	var resultObj object.Object
	switch ope.Type {
	case token.TT_PLUS:
		resultObj = &object.NumberObj{Value: float64(leftObj.Value + rightObj.Value)}
	case token.TT_MINUS:
		resultObj = &object.NumberObj{Value: float64(leftObj.Value - rightObj.Value)}
	case token.TT_MUL:
		resultObj = &object.NumberObj{Value: float64(leftObj.Value * rightObj.Value)}
	case token.TT_DIV:
		var rightValue = rightObj.Value
		if rightValue < 0 {
			return object.NewError("Division by zero.")
		}
		resultObj = &object.NumberObj{Value: float64(leftObj.Value / rightObj.Value)}
	case token.TT_LESS:
		if leftObj.Value < rightObj.Value {
			return TRUE
		} else {
			return FALSE
		}
	case token.TT_LESS_EQ:
		if leftObj.Value <= rightObj.Value {
			return TRUE
		} else {
			return FALSE
		}
	case token.TT_GREATER:
		if leftObj.Value > rightObj.Value {
			return TRUE
		} else {
			return FALSE
		}
	case token.TT_GREATER_EQ:
		if leftObj.Value >= rightObj.Value {
			return TRUE
		} else {
			return FALSE
		}
	case token.TT_EQUAL:
		if leftObj.Value == rightObj.Value {
			return TRUE
		} else {
			return FALSE
		}
	case token.TT_NOT_EQ:
		if leftObj.Value != rightObj.Value {
			return TRUE
		} else {
			return FALSE
		}
	default:
		resultObj = object.NewError(fmt.Sprintf("Invalid operator: %s", ope.Literal))
	}

	return resultObj
}

// evalBinaryString
func evalBinaryString(leftObj *object.StringObj, ope token.Token, rightObj *object.StringObj) object.Object {
	var resultObj object.Object
	switch ope.Type {
	case token.TT_PLUS:
		resultObj = &object.StringObj{Value: string(leftObj.Value + rightObj.Value)}
	default:
		resultObj = object.NewError(fmt.Sprintf("Invalid operator: %s\n", ope.Literal))
	}

	return resultObj
}

// evalBooleanNode
func evalBooleanNode(booleanNode *ast.BooleanNode) object.Object {
	if booleanNode.Value {
		return TRUE
	} else {
		return FALSE
	}
}

// evalIdentifier
func evalIdentifier(identifierNode *ast.IdentifierNode, env *object.Environment) object.Object {
	if valueObj, ok := env.Get(identifierNode.Value); ok {
		return valueObj
	}
	// check for builtin function
	if builtinObj, ok := builtins[identifierNode.Value]; ok {
		return builtinObj
	}
	// definetely is not registered.
	return object.NewError(fmt.Sprintf("identifier not found: '%s'", identifierNode.Value))
}

// evalFunctionLiteralNode
func evalFunctionLiteralNode(functionLiteralNode *ast.FunctionLiteralNode, env *object.Environment) object.Object {
	return &object.FunctionObj{
		Parameters: functionLiteralNode.Parameters,
		Body:       functionLiteralNode.Body,
		Env:        env,
	}
}

File: Cpp/src/evaluator.cpp
//
// Created by irwin on 13/05/2021.
//
#include "../header/evaluator.h"

namespace corny {
    // check whether passed object is null or ErrorObj type.
    bool Evaluator::isError(Object *obj) {
        return obj != nullptr && obj->type == OBJ_ERROR;
    }
    // evalStatements
    Object* Evaluator::evalStatements(std::vector<Node*> statements, Environment *env) {
        Object* resultObj;
        for (auto statement : statements) {
            resultObj = eval(statement, env);
            gc.mark(resultObj); // set the mark to true.
            // if it is a OBJ_RETURN then mark the value
            if (resultObj->type == OBJ_RETURN) {
                gc.mark(((ReturnObj*)resultObj)->value);
            }
            // now increase the gcCounter
            gcCounter += 1;
            // check for the limit and sweep
            if (gcCounter == gcMaxObjects) {
                gc.mark(env); // mark the symbol table.
                gc.sweep(); // start sweeping all objects.
                gcCounter = 0; // restart the objects counter.
            }

            if (resultObj->type == OBJ_RETURN || resultObj->type == OBJ_ERROR) {
                return resultObj;
            }
        }

        return resultObj;
    }
    // recursively evaluates the current node.
    Object* Evaluator::eval(Node *node, Environment *env) {
        NodeType type = node->type;
        switch(type) {
            case NT_PROGRAM:
                return evalProgram((ProgramNode*)node, env);
            case NT_BLOCK:
                return evalBlock((BlockNode*)node, env);
            case NT_BOOLEAN:
                return (((BooleanNode*)node)->value) ? TRUE : FALSE;
            case NT_NUMBER:
            {
                NumberObj* numberObj = new NumberObj(((NumberNode*)node)->value);
                gc.add(numberObj);
                return numberObj;
            }
            case NT_NULL:
                return NIL;
            case NT_UNARY:
                return evalUnaryExpression((UnaryNode*)node, env);
            case NT_BINARY:
                return evalBinaryExpression((BinOpNode*)node, env);
            case NT_STRING:
            {
                StringObj* stringObj = new StringObj(((StringNode*)node)->value);
                gc.add(stringObj);
                return stringObj;
            }
            case NT_LET:
                return evalLet((LetNode*)node, env);
            case NT_RETURN:
                return evalReturn((ReturnNode*)node, env);
            case NT_IDENT:
                return evalIdentifier((IdentNode*)node, env);
            case NT_ARRAY:
                return evalArrayLiteral((ArrayNode*)node, env);
            case NT_HASH:
                return evalHashLiteral((HashNode*)node, env);
            case NT_CALL:
                return evalCallExpr((CallExprNode*)node, env);
            case NT_FUNCTION:
                return evalFunctionLiteral((FunctionNode*)node, env);
            case NT_IF:
                return evalIfExpression((IfNode*)node, env);
            default:
                return new ErrorObj("Unknown Node type.");
        }
    }
    // evalProgram
    Object* Evaluator::evalProgram(ProgramNode *programNode, Environment *env) {
        Object* resultObj = evalStatements(programNode->statements, env);
        // unwrap the return value
        if (resultObj->type == OBJ_RETURN) {
            return ((ReturnObj*)resultObj)->value;
        }
        return resultObj;
    }
    // evalBlock
    Object* Evaluator::evalBlock(BlockNode *blockNode, Environment *env) {
        return evalStatements(blockNode->statements, env);
    }
    // evalLet
    Object* Evaluator::evalLet(LetNode *letNode, Environment *env) {
        // evaluate the value property
        Object* valueObj = eval(letNode->value, env);
        if (isError(valueObj)) return valueObj;
        // register the symbol
        env->set(letNode->ident->value.literal, valueObj);

        return valueObj;
    }
    // evalReturn
    Object* Evaluator::evalReturn(ReturnNode *returnNode, Environment *env) {
        Object* valueObj = eval(returnNode->value, env);
        if (isError(valueObj)) return valueObj;

        return new ReturnObj(valueObj);
    }
    // evalIfExpression
    Object* Evaluator::evalIfExpression(IfNode *ifNode, Environment *env) {
        Object* conditionObj = eval(ifNode->condition, env);
        if (isError(conditionObj)) return conditionObj;
        if (conditionObj->type != OBJ_BOOLEAN) return new ErrorObj("Invalid data type for if condition");
        // evaluate if or else based on condition value.
        if (((BooleanObj*)conditionObj)->value == true) {
            return eval(ifNode->consequence, env);
        } else {
            if (ifNode->alternative != nullptr) {
                return eval(ifNode->alternative, env);
            } else {
                return NIL;
            }
        }
    }
    // evalUnaryExpression
    Object* Evaluator::evalUnaryExpression(UnaryNode *unaryNode, Environment *env) {
        Object* resultObj;
        Object* rightObj = eval(unaryNode->left, env);
        if (isError(rightObj)) return rightObj;
        // check the token operator
        switch (unaryNode->opToken.type) {
            case TT_MINUS:
                if (rightObj->type != OBJ_NUMBER) return new ErrorObj("Invalid data type");
                resultObj = new NumberObj(((NumberObj*)rightObj)->value * -1);
                break;
            case TT_NOT:
                if (rightObj->type != OBJ_BOOLEAN) return new ErrorObj("Invalid data type");
                resultObj = (((BooleanObj*)rightObj)->value == true) ? FALSE : TRUE;
                break;
            default:
                return new ErrorObj("Invalid operator: " + unaryNode->opToken.literal);
        }
        gc.add(resultObj);
        return resultObj;
    }
    // evalBinaryExpression
    Object* Evaluator::evalBinaryExpression(BinOpNode *binOpNode, Environment *env) {
        if (binOpNode->opToken.type == TT_AND || binOpNode->opToken.type == TT_OR) {
            return evalLogicalExpression(binOpNode, env);
        }
        // we need to know the both operands types
        Object* leftObj = eval(binOpNode->left, env);
        if (isError(leftObj)) return leftObj;
        Object* rightObj = eval(binOpNode->right, env);
        if (isError(rightObj)) return rightObj;
        // now based on their types we perform the correct operations
        if (leftObj->type == OBJ_STRING && rightObj->type == OBJ_STRING) {
            return evalBinaryString(leftObj, binOpNode->opToken.type, rightObj);
        }
        if (leftObj->type == OBJ_NUMBER && rightObj->type == OBJ_NUMBER) {
            return evalBinaryInteger(leftObj, binOpNode->opToken.type, rightObj);
        }
        if (leftObj->type == OBJ_BOOLEAN && rightObj->type == OBJ_NUMBER) {
            return evalBinaryBoolean(leftObj, binOpNode->opToken.type, rightObj);
        }
    }
    // evalLogicalExpression
    Object* Evaluator::evalLogicalExpression(BinOpNode *binOpNode, Environment *env) {
        // evaluate the left hand operator
        Object* leftObj = eval(binOpNode->left, env);
        if (isError(leftObj)) return leftObj;
        if (leftObj->type != OBJ_BOOLEAN) return new ErrorObj("Invalid left hand type operand");
        if (binOpNode->opToken.type == TT_AND) {
            // if leftObj is false then there's nothing else to do.
            if (((BooleanObj*)leftObj)->value == true) return FALSE;
            // otherwise we need to evaluate the right hand operator
            Object* rightObj = eval(binOpNode->right, env);
            if (isError(rightObj)) return rightObj;
            if (rightObj->type != OBJ_BOOLEAN) return new ErrorObj("Invalid right hand type operand");

            return (((BooleanObj*)leftObj)->value == true) ? TRUE : FALSE;
        }
        else if (binOpNode->opToken.type == TT_OR){
            // if leftObj is true then there's nothing else to do.
            if (((BooleanObj*)leftObj)->value == true) return TRUE;
            // otherwise we need to evaluate the right hand operator
            Object* rightObj = eval(binOpNode->right, env);
            if (isError(rightObj)) return rightObj;
            if (rightObj->type != OBJ_BOOLEAN) return new ErrorObj("Invalid right hand type operand");

            return (((BooleanObj*)leftObj)->value == true) ? TRUE : FALSE;
        }
        return new ErrorObj("Invalid operator for logical operation: " + binOpNode->opToken.literal);
    }
    // evalBinaryString
    Object* Evaluator::evalBinaryString(Object* leftObj, TokenType type, Object* rightObj) {
        Object* resultObj;
        switch (type) {
            case TT_PLUS:
                resultObj = new StringObj(((StringObj*)leftObj)->value + ((StringObj*)rightObj)->value);
                break;
            default:
                return new ErrorObj("Invalid operator");
        }
        gc.add(resultObj);
        return resultObj;
    }
    // evalBinaryInteger
    Object* Evaluator::evalBinaryInteger(Object* leftObj, TokenType type, Object* rightObj) {
        Object* resultObj;
        switch (type) {
            case TT_PLUS:
                resultObj = new NumberObj(((NumberObj*)leftObj)->value + ((NumberObj*)rightObj)->value);
                break;
            case TT_MINUS:
                resultObj = new NumberObj(((NumberObj*)leftObj)->value - ((NumberObj*)rightObj)->value);
                break;
            case TT_MUL:
                resultObj = new NumberObj(((NumberObj*)leftObj)->value * ((NumberObj*)rightObj)->value);
                break;
            case TT_DIV:
            {
                double rightVal = ((NumberObj*)rightObj)->value;
                if (rightVal <= 0) return new ErrorObj("Division by zero.");
                resultObj = new NumberObj(((NumberObj*)leftObj)->value / rightVal);
                break;
            }
            case TT_LESS:
                return (((NumberObj*)leftObj)->value < ((NumberObj*)rightObj)->value) ? TRUE : FALSE;
            case TT_GREATER:
                return (((NumberObj*)leftObj)->value > ((NumberObj*)rightObj)->value) ? TRUE : FALSE;
            case TT_LESS_EQ:
                return (((NumberObj*)leftObj)->value <= ((NumberObj*)rightObj)->value) ? TRUE : FALSE;
            case TT_GREATER_EQ:
                return (((NumberObj*)leftObj)->value >= ((NumberObj*)rightObj)->value) ? TRUE : FALSE;
            case TT_EQUAL:
                return (((NumberObj*)leftObj)->value == ((NumberObj*)rightObj)->value) ? TRUE : FALSE;
            case TT_NOT_EQ:
                return (((NumberObj*)leftObj)->value != ((NumberObj*)rightObj)->value) ? TRUE : FALSE;
            default:
                return new ErrorObj("Invalid operator.");
        }
        gc.add(resultObj); // add in GC
        return resultObj;
    }
    // evalBinaryBoolean
    Object* Evaluator::evalBinaryBoolean(Object* leftObj, TokenType type, Object* rightObj) {
        // this is the trick: transform boolean to int and call binary integer operations.
        NumberObj *leftNumber = new NumberObj((((BooleanObj*)leftObj)->value == true) ? 1 : 0);
        NumberObj *rightNumber = new NumberObj((((BooleanObj*)rightObj)->value == true) ? 1 : 0);
        return evalBinaryInteger(leftNumber, type, rightNumber);
    }
    // evalIdentifier
    Object* Evaluator::evalIdentifier(IdentNode *identNode, Environment *env) {
        std::string identifier = identNode->value.literal;
        Object *valueObj = env->get(identifier);
        if (valueObj == nullptr) {
            return new ErrorObj("variable not defined: " + identifier);
        }
        return valueObj;
    }
    // evalCallExpression
    Object* Evaluator::evalCallExpr(CallExprNode *callExprNode, Environment *env) {
        // this node can contain function call, array call or hash table call.
        // 1. get the object of the callee
        Object* calleeObj = eval(callExprNode->callee, env);
        if (isError(calleeObj)) return calleeObj;
        // 2. evaluate the arguments
        std::vector<Object*> arguments;
        if (callExprNode->arguments.size() > 0) {
            Object* resultObj;
            for (auto argument : callExprNode->arguments) {
                resultObj = eval(argument, env);
                if (isError(resultObj)) return resultObj;
                arguments.emplace_back(resultObj);
            }
        }
        // 3. check the callee type
        ObjType type = calleeObj->type;
        switch (type) {
            case OBJ_FUNCTION:
                return evalFunction((FunctionObj*)calleeObj, arguments);
            case OBJ_ARRAY:
                return evalArrayAccess((ArrayObj*)calleeObj, arguments);
            case OBJ_HASH:
                return evalHashAccess((HashObj*)calleeObj, arguments);
            case OBJ_STRING:
                return evalStringAccess((StringObj*)calleeObj, arguments);
            default:
                return new ErrorObj("Invalid callable object.");
        }

    }
    // evalFunctionLiteral
    Object* Evaluator::evalFunctionLiteral(FunctionNode *functionNode, Environment *env) {
        FunctionObj *functionObj = new FunctionObj();
        functionObj->env = env;
        functionObj->parameters = functionNode->parameters;
        functionObj->body = functionNode->body;
        gc.add(functionObj); // add in garbage collector
        return functionObj;
    }
    // evalArrayLiteral
    Object* Evaluator::evalArrayLiteral(ArrayNode *arrayNode, Environment *env) {
        ArrayObj *arrayObj = new ArrayObj();
        // evaluate the elements of the array
        if (arrayNode->elements.size() > 0) {
            Object* resultObj;
            for (auto element : arrayNode->elements) {
                resultObj = eval(element, env);
                if (isError(resultObj)) return resultObj;
                // push the element into the array.
                arrayObj->elements.emplace_back(resultObj);
            }
        }
        gc.add(arrayObj);
        return arrayObj;
    }
    // evalHashLiteral
    Object* Evaluator::evalHashLiteral(HashNode *hashNode, Environment *env) {
        HashObj* hashObj = new HashObj();
        int index = -1;
        Object* keyObj, *valueObj;
        // loop through keys and their values
        for (auto keyNode : hashNode->keys) {
            index += 1;
            keyObj = eval(keyNode, env);
            if (isError(keyObj)) return keyObj;
            // validate the OBJ_STRING data type
            if (keyObj->type != OBJ_STRING) return new ErrorObj("Invalid data type for key");
            // evaluate the value
            valueObj = eval(hashNode->values.at(index), env);
            if (isError(valueObj)) return valueObj;
            // save the key-value in data type
            hashObj->elements[((StringObj*)keyObj)->value] = valueObj;
        }
        gc.add(hashObj);
        return hashObj;
    }
    // evalFunction
    Object* Evaluator::evalFunction(FunctionObj *functionObj, std::vector<Object *> arguments) {
        // 1. create new environment for the function
        Environment* newEnv = new Environment();
        newEnv->outer = functionObj->env; // enclose environment

        // 2. check for function arity.
        int numArgs = arguments.size();
        int numParams = functionObj->parameters.size();
        if (numArgs != numParams) return new ErrorObj("Unexpected arguments, got: " + std::to_string(numArgs) + " want: " + std::to_string(numParams));

        // 3. fill the new environment with arguments
        for (int i = 0; i < numParams; i++) {
            newEnv->set(functionObj->parameters.at(i)->value.literal, arguments.at(i));
        }
        // 4. execute the function with new environment
        Object *resultObj = eval(functionObj->body, newEnv);
        if (isError(resultObj)) return resultObj;
        // 5. check for return
        if (resultObj->type == OBJ_RETURN) return ((ReturnObj*)resultObj)->value;

        return resultObj;
    }
    // evalArrayAccess
    Object* Evaluator::evalArrayAccess(ArrayObj *arrayObj, std::vector<Object *> arguments) {
        Object* indexObj = arguments.at(0);
        if (indexObj->type != OBJ_NUMBER) return new ErrorObj("Invalid subscript reference");
        // check for out of bounds
        int index = ((NumberObj*)indexObj)->value;
        if (index < 0 || index >= arrayObj->elements.size()) {
            return new ErrorObj("Index out of bounds");
        }

        return arrayObj->elements[index];
    }
    // evalHashAccess
    Object* Evaluator::evalHashAccess(HashObj *hashObj, std::vector<Object *> arguments) {
        Object* indexObj = arguments.at(0);
        if (indexObj->type != OBJ_STRING) return new ErrorObj("Invalid subscript reference");
        std::string key = ((StringObj*)indexObj)->value;
        if (hashObj->elements.find(key) != hashObj->elements.end()) {
            return hashObj->elements[key];
        }
        return NIL;
    }
    // stringAccess
    Object* Evaluator::evalStringAccess(StringObj *stringObj, std::vector<Object *> arguments) {
        Object* indexObj = arguments.at(0);
        if (indexObj->type != OBJ_NUMBER) return new ErrorObj("Invalid subscript reference");
        int index = ((NumberObj*)indexObj)->value;
        // check for out of bounds
        if (index < 0 || index > stringObj->value.length()) {
            return new ErrorObj("Index out of bounds");
        }

        return new StringObj(std::string(1, stringObj->value[index]));
    }
}